docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+1. VMEM usage: The bigger our blocks, the more VMEM we use. With large enough blocks, we will run out.
+2. Pipeline bubbles: The larger our blocks are relative to the matrix size, the fewer loop iterations we will have in our pipeline. This will make the size of the bubbles at the beginning and end of the pipeline larger relative to the total pipeline and this overhead can be nontrivial.
+
+Getting good matrix multiplication performance in Pallas boils down to picking good block sizes to balance this optimization problem. In practice, we often sweep over a large set of candidate block sizes, profile the kernel, and pick the best one.
+
+For now, let's do some very simple timing experiments. We'll use `timeit` to measure the amount of time running each kernel takes. Note that this is a upper bound on the actual runtime of the kernel since we are measuring Python dispatch and other overheads using `timeit`. We'll compute the amount of FLOP/s we obtained this way and compute the percentage utilization we get compared to what the chip offers and we'll use some reasonable block sizes to verify our intuition.
+
+```{code-cell} ipython3
+:id: RjU3sSTUWzIk
+:outputId: 02b5793e-1ff3-41f4-ab45-4cf1393885ba
+
+import timeit
+
+def benchmark(f, ntrials: int = 100):
+  def run(*args, **kwargs):
+    # Compile function first
+    jax.block_until_ready(f(*args, **kwargs))
+    # Time function
+    result = timeit.timeit(lambda: jax.block_until_ready(f(*args, **kwargs)),
+                           number=ntrials)
+    time = result / ntrials
+    # print(f"Time: {time}")
+    return time
+  return run
+
+def analyze_matmul(m: int, k: int, n: int, dtype: np.dtype,
+                   mm_func):
+  x = jnp.ones((m, k), dtype=dtype)
+  y = jnp.ones((k, n), dtype=dtype)
+  time = benchmark(mm_func)(x, y)
+  print(f"----- {m} x {k} x {n} -----")
+  print("Matmul time: ", time)
+  mm_flops = matmul_flops(m, k, n) / time
+  print("Matmul FLOP/s: ", mm_flops)
+  print(f"FLOPs utilization: {mm_flops / v5e_bf16_flops * 100:.4f}%")
+  print()
+
+print("================bm=128, bk=128, bn=128===================")
+mm = functools.partial(matmul, bm=128, bk=128, bn=128)
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+
+print("================bm=512, bk=1024, bn=1024===================")
+mm = functools.partial(matmul, bm=512, bk=1024, bn=1024)
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+```
+
++++ {"id": "mg1GMqcVan70"}
+
+Bigger block sizes help a lot! We get pretty good utilization (80-90%) in the bigger matmuls, but the smallest matmul seems pretty hard to get good performance with.
+
+Let's compare this with XLA's matmuls. We don't expect Pallas to do better than XLA because XLA is *very* good at generating matmuls but hopefully we are close.
```

Comment:
hopefully we are close -> we should be close.

+ add that with more careful block size tuning you also *can* reach XLA performance. This sentence sounds as if XLA can do tricks that Pallas can't